src/hooks/use-debounced-block.ts

```diff
@@ -0,0 +1,61 @@
+import { useCallback, useEffect, useState } from "react";
+
+import { ConnectedProvider } from "src/domain";
+import { AsyncTask } from "src/utils/types";
+
+type DebouncedBlockCallback = (blockNumber: number) => void;
+
+const DEBOUNCE_TIME_IN_MS = 750;
+
+const useDebouncedBlock = (connectedProvider: AsyncTask<ConnectedProvider, string>) => {
+  const [listener, setListener] = useState<DebouncedBlockCallback>();
+  const [blockNumber, setBlockNumber] = useState<number>();
+
+  const addBlockListener = useCallback(
+    (cb: DebouncedBlockCallback) => {
+      if (!listener) {
+        setListener(() => cb);
+      }
+    },
+    [listener]
+  );
+
+  useEffect(() => {
+    const timeoutId = setTimeout(() => {
+      if (listener && blockNumber) {
+        listener(blockNumber);
+      }
+    }, DEBOUNCE_TIME_IN_MS);
+
+    return () => {
+      clearTimeout(timeoutId);
+    };
+  }, [blockNumber, listener]);
+
+  useEffect(() => {
+    if (listener) {
+      const onBlock = (newBlockNumber: number) => {
+        if (!blockNumber || newBlockNumber > blockNumber) {
+          setBlockNumber(newBlockNumber);
+        }
+      };
+
+      if (connectedProvider.status === "successful") {
+        void connectedProvider.data.provider.getBlockNumber().then((currentBlockNumber) => {
+          onBlock(currentBlockNumber);
+          connectedProvider.data.provider.on("block", onBlock);
+        });
+      }
+
+      return () => {
+        if (connectedProvider.status === "successful") {
+          connectedProvider.data.provider.off("block", onBlock);
+        }
+      };
+    }
+  }, [blockNumber, connectedProvider, listener]);
+
+  return addBlockListener;
+};
+
+export default useDebouncedBlock;
```

Comment:
I feel this hook is over-complicated, partly because it couples two concepts that should remain abstracted:

* Debouncing
* Managing the subscription to new blocks

I'd like to consider an alternative based on 2 simpler hooks:

A generic, multipurpose, reusable `use-debounce` hook:

```typescript
import { useEffect, useState } from 'react'

export function useDebounce<T>(value: T, delay: number): T {
  const [debouncedValue, setDebouncedValue] = useState<T>(value)

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value)
    }, delay)

    return () => {
      clearTimeout(handler)
    }
  }, [value, delay])

  return debouncedValue
}
```

A specific `use-debounce-block` that manages the subscription and debounces the blocks using the other hook :

```typescript
import { useEffect, useState } from "react";

import { useDebounce } from "src/hooks/use-debounce";
import { AsyncTask, isAsyncTaskDataAvailable } from "src/utils/types";
import { ConnectedProvider } from "src/domain";

const DEBOUNCE_TIME_IN_MS = 750;

const useDebouncedBlock = (connectedProvider: AsyncTask<ConnectedProvider, string>) => {
  const [blockNumber, setBlockNumber] = useState<number>();

  useEffect(() => {
    if (isAsyncTaskDataAvailable(connectedProvider)) {
      void connectedProvider.data.provider
        .getBlockNumber()
        .then(setBlockNumber)
        .then(() => {
          connectedProvider.data.provider.on("block", setBlockNumber);
        });

      return () => {
        connectedProvider.data.provider.off("block", setBlockNumber);
      };
    }
  }, [connectedProvider]);

  return useDebounce(blockNumber, DEBOUNCE_TIME_IN_MS);
};

export default useDebouncedBlock;
```

It just returns the new debounced block when it arrives. It can be used like this:

```typescript
useEffect(() => {
  if (newBlock) {
    console.log("A new block has been mined:", newBlock);
  }
}, [newBlock]);
```

Which results in:

<img src="https://user-images.githubusercontent.com/15896177/198000224-299a68b5-544d-45bb-90c6-87d4f3dc8a80.gif" width="300px" />

Do you see any drawbacks to this approach?

Some benefits I see:

* Simpler code to follow/maintain with encapsulated functionality.
* No listeners/callbacks need to be passed to the block debouncer.
* We get a generic debouncer that can be used for anything that requires debouncing (like an input).

Reply:
I agree, this is more `hooks` base while i'm more used to node `EventEmitter` like patterns, in this case this will match React architecture better.